guides/micronaut-openapi-generator-server/micronaut-openapi-generator-server.adoc

```diff
@@ -0,0 +1,550 @@
+
+=== Testing the Controller
+
+We will write tests for the two paths of `BookController`.
+
+Create a `BooksControllerTest` with the following contents:
+
+test:controller/BooksControllerTest[]
+
+callout:micronaut-test[1]
+callout:http-client[2]
+callout:http-request[3]
+<4> Verify that addition of book info was successful by checking the status code.
+callout:binding-json-array[5]
+callout:body-method[6]
+<7> Verify that there are exactly two books with `"Guide"` substring in title.
+
+common:testApp-noheader.adoc[]
+
+All the tests should run successfully.
+
+common:graal-with-plugins.adoc[]
+
+== Next steps
+
+=== Learn More
+
+Read OpenAPI and Micronaut documentation and guides:
+
+* https://www.openapis.org[OpenAPI]
+* Definition files generation from annotated controllers with link:https://micronaut-projects.github.io/micronaut-openapi/latest/guide[Micronaut OpenAPI]
+* link:https://micronaut-projects.github.io/micronaut-data/latest/guide[Micronaut Data]
+
+=== Add Security
+
+We could have defined our security requirements by adding a security schema to the `library-definition.yml` file.
+For example, we will add HTTP Basic authentication:
+
+[source,yaml]
+----
+paths:
+  /search:
+    # ... #
+  /add:
+    post:
+      # ... #
+      security:
+        - MyBasicAuth: [] # <2>
+components:
+  schemas:
+    # ... #
+  securitySchemes:
+    MyBasicAuth: # <1>
+      type: http
+      scheme: basic
+----
+<1> Define a security schema inside the `components/securitySchemes`. We want to use Basic auth for authentication.
+<2> Add the schema to the paths that you want to secure. In this case, we want to restrict access to
+adding books into our library.
+
+NOTE: You can read more about describing various authentication in the
+link:https://swagger.io/docs/specification/authentication/["Authentication and Authorization" OpenAPI guide].
+
+The generator will then annotate such endpoints with the
+link:https://micronaut-projects.github.io/micronaut-security/latest/guide/#secured[Secured] annotation accordingly:
+
+[source,java]
+----
+@Secured(SecurityRule.IS_AUTHENTICATED)
+public void addBook( /* ... */ ){ /* ... */ }
+----
+
+You will then need to implement an
+link:https://micronaut-projects.github.io/micronaut-security/latest/guide/#authenticationProviders[AuthenticationProvider]
+that satisfies your needs. If you want to finish implementing the basic authentication, continue to the
+link:https://guides.micronaut.io/latest/micronaut-security-basicauth.html[Micronaut Basic Auth guide] and replicate
+steps to create the `AuthenticationProvider` and appropriate tests.
+
+NOTE: You can also read link:https://micronaut-projects.github.io/micronaut-security/latest/guide/[Micronaut Security documentation]
+or link:https://micronaut.io/guides[Micronaut guides] about security to learn more about
+the supported Authorization strategies.
+
+=== Generate Micronaut Client
+
```

Comment:
We could add this info:
```
To specify the server URL to be consumed by OpenAPI client generators, use the link:servers list property[https://swagger.io/specification/#server-object] of the OpenAPI specification.
```

Reply:
Currently it can be changed by configuring the `artifactId` property. But I agree that we should improve this. We should use micronaut configuration via client id instead: https://docs.micronaut.io/latest/guide/#_client_specific_configuration (`micronaut.http.services.client-id.urls` for `@Client("[client-id]")`.